slave-base/contrib/bin/run-jnlp-client

```diff
@@ -65,57 +63,88 @@ if [[ $# -gt 1 ]]; then
+  if [[ -z "${JAVA_TOOL_OPTIONS}" ]]; then
+    # these options will automatically be picked up by any child JVM process
+    # but can be overridden on that process' command line.
+    JAVA_TOOL_OPTIONS="-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -Dsun.zip.disableMemoryMapping=true"
+    export JAVA_TOOL_OPTIONS
+  fi
 
-    if [ -z $CONTAINER_HEAP_PERCENT ]; then
+  CONTAINER_MEMORY_IN_BYTES=`cat /sys/fs/cgroup/memory/memory.limit_in_bytes`
+  # assume k8s/docker memory limit was set if memory.limit_in_bytes < 1TiB
+  if [[ "${CONTAINER_MEMORY_IN_BYTES}" -lt $((2**40)) ]]; then
+    # set this JVM's -Xmx and -Xms if not set already (not propagated to any
+    # child JVMs).  -Xmx can be calculated as a percentage, capped to a maximum,
+    # or specified straight.  -Xms can be calculated as a percentage or
+    # specified straight.  For the JNLP slave by default we specify -Xmx of 50%,
+    # uncapped; -Xms unspecified (JVM default is 1/64 of -Xmx).
+    CONTAINER_MEMORY_IN_MB=$((CONTAINER_MEMORY_IN_BYTES/2**20))
+
+    if [[ -z "$CONTAINER_HEAP_PERCENT" ]]; then
         CONTAINER_HEAP_PERCENT=0.50
     fi
 
-    CONTAINER_MEMORY_IN_MB=$((${CONTAINER_MEMORY_IN_BYTES}/1024**2))
-    #if machine has 4GB or less, meaning max heap of 2GB given current default, force use of 32bit to save space unless user
-    #specifically want to force 64bit
-    HEAP_LIMIT_FOR_32BIT=$((2**32-1))
-    HEAP_LIMIT_FOR_32BIT_IN_MB=$((${HEAP_LIMIT_FOR_32BIT}/1024**2))
     CONTAINER_HEAP_MAX=$(echo "${CONTAINER_MEMORY_IN_MB} ${CONTAINER_HEAP_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
-    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH && "${CONTAINER_HEAP_MAX}" -lt "${HEAP_LIMIT_FOR_32BIT_IN_MB}"  ]]; then
-      echo "max heap in MB is ${CONTAINER_HEAP_MAX} and 64 bit was not explicitly set so using 32 bit Java"
-      alternatives --set java $JVMPath32bit
-      export MALLOC_ARENA_MAX=1
+    if [[ $CONTAINER_HEAP_LIMIT_IN_MB && $CONTAINER_HEAP_MAX -gt $CONTAINER_HEAP_LIMIT_IN_MB ]]; then
+      CONTAINER_HEAP_MAX=$CONTAINER_HEAP_LIMIT_IN_MB
+    fi
+    if [[ -z "$JAVA_MAX_HEAP_PARAM" ]]; then
+      JAVA_MAX_HEAP_PARAM="-Xmx${CONTAINER_HEAP_MAX}m"
     fi
 
-    JAVA_MAX_HEAP_PARAM="-Xmx${CONTAINER_HEAP_MAX}m"
-    if [ -z $CONTAINER_INITIAL_PERCENT ]; then
-      CONTAINER_INITIAL_PERCENT=0.07
+    if [[ "$CONTAINER_INITIAL_PERCENT" ]]; then
+      CONTAINER_INITIAL_HEAP=$(echo "${CONTAINER_HEAP_MAX} ${CONTAINER_INITIAL_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
+      if [[ -z "$JAVA_INITIAL_HEAP_PARAM" ]]; then
+        JAVA_INITIAL_HEAP_PARAM="-Xms${CONTAINER_INITIAL_HEAP}m"
+      fi
+    fi
+
+    # if the calculated max heap is < 2GiB and OPENSHIFT_JENKINS_JVM_ARCH is not
+    # set, use 32 bit JVM for space efficiency
+    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH ]]; then
+      if [[ "${CONTAINER_HEAP_MAX}" -gt 2048 ]]; then
+        echo "max heap in MB is ${CONTAINER_HEAP_MAX}, using 64 bit Java"
```

Comment:
While you're cleaning things up, both of these messages(here and the 32 bit one below) should say "OPENSHIFT_JENKINS_JVM_ARCH was not set and max heap in MB is X.  Using Y bit Java"